p2p/gossip.go

```diff
@@ -0,0 +1,66 @@
+// Copyright 2019 ChainSafe Systems (ON) Corp.
+// This file is part of gossamer.
+//
+// The gossamer library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The gossamer library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more detailg.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the gossamer library. If not, see <http://www.gnu.org/licenses/>.
+
+package p2p
+
+import (
+	log "github.com/ChainSafe/log15"
+	"github.com/libp2p/go-libp2p-core/network"
+)
+
+// gossip submodule
+type gossip struct {
+	host        *host
+	hasGossiped map[string]bool
+}
+
+// newGossip creates a new gossip instance from the host
+func newGossip(host *host) (g *gossip, err error) {
+	g = &gossip{
+		host:        host,
+		hasGossiped: make(map[string]bool),
+	}
+	return g, err
+}
+
+// handleMessage gossips messages that have not already been gossiped
+func (g *gossip) handleMessage(stream network.Stream, msg Message) {
+
+	// check if message has been gossiped
+	if !g.hasGossiped[msg.Id()] {
+
+		// broadcast message to peers if message has not been gossiped
+		g.sendMessage(stream, msg)
+
+		// update message to gossiped
+		g.hasGossiped[msg.Id()] = true
+
+	}
+}
+
+// sendMessage broadcasts the message to connected peers
+func (g *gossip) sendMessage(stream network.Stream, msg Message) {
+
+	// loop through connected peers
+	for _, peer := range g.host.peers() {
+
+		// send message to each connected peer
+		err := g.host.send(peer, msg)
+		if err != nil {
+			log.Error("Failed to send message during gossip", "err", err)
+		}
+	}
+}
```

Comment:
I think this can replaced by `host.Broadcast`

Reply:
Yup, it can. I was not sure whether I should keep it separate given gossip might evolve.

Reply:
Should we keep it separate it preparation? It's also nice to have the error include "gossip" so that it is more specific for debugging and the user.

Reply:
I think this was also a result of the reference to a send method in the issue.

Reply:
Ok, updated. Simplified to what the `gossip` submodule breaks down to at the moment. We can add when we know need to.